sql/glam_etl/org_mozilla_fenix__histogram_bucket_counts_v1/query.sql

```diff
@@ -80,7 +80,41 @@ RETURNS ARRAY<
   )
 );
 
-WITH normalized_histograms AS (
+WITH
+-- Cross join with the attribute combinations to reduce the query complexity
+-- with respect to the number of operations. A table with n rows cross joined
+-- with a combination of m attributes will generate a new table with n*m rows.
+-- The glob ("*") symbol can be understood as selecting all of values belonging
+-- to that group.
+static_combos AS (
+  SELECT
+    combos.*
+  FROM
+    UNNEST(
+      ARRAY<STRUCT<ping_type STRING, os STRING, app_build_id STRING>>[
+        (NULL, NULL, NULL),
+        (NULL, NULL, "*"),
+        (NULL, "*", NULL),
+        ("*", NULL, NULL),
+        (NULL, "*", "*"),
+        ("*", NULL, "*"),
+        ("*", "*", NULL),
+        ("*", "*", "*")
+      ]
+    ) AS combos
+),
+all_combos AS (
+  SELECT
+    table.* EXCEPT (ping_type, os, app_build_id),
+    COALESCE(combo.ping_type, table.ping_type) AS ping_type,
+    COALESCE(combo.os, table.os) AS os,
+    COALESCE(combo.app_build_id, table.app_build_id) AS app_build_id
+  FROM
+    glam_etl.org_mozilla_fenix__clients_histogram_aggregates_v1 table
+  CROSS JOIN
+    static_combos combo
+),
+normalized_histograms AS (
```

Comment:
Before applying normalization, there needs to be a `group by` including `client_id` so that the clients that have multiple rows for `*` build_id are grouped into 1 row to represent them and that is the row that gets normalized

Reply:
This is to catch cases where a null build id is being reported?

Reply:
My understanding is that the source table (clients histogram aggregates) should only have a single row per dimension. Null build ids should be invalid (though I notice there is no filter in the base queries for clients daily). Do you have an example where a single row could have multiple `*` in the cross joined table?

Reply:
Sorry, my explanation wasn't very good. You handled it correctly with your new changes. I just meant that a single client could have multiple builds so when we group over all build ids they need to be deduped which you have done here.